### PR TITLE
fix(connection/actions.native): dispatch missing CONFERENCE_LEFT

### PR DESCRIPTION
**Pull request by paweldomas**
Currently it's possible for lib-jitsi-meet to not dispatch CONFERENCE_LEFT event in case of a ChatRoom.leave() timeout (rejected promise). There's a logic there to dispatch participant left events,
but no CONFERENCE_LEFT.

### CODE REVIEW

**Comment by paweldomas** [comment]
The flow is that room.leave() promise[1] is rejected with a timeout[2] and the catch block does not emit CONFERENCE_LEFT.
[1]: https://github.com/jitsi/lib-jitsi-meet/blob/master/JitsiConference.js#L421
[2]: https://github.com/jitsi/lib-jitsi-meet/blob/master/modules/xmpp/ChatRoom.js#L1392

We may or we may not want this fix in I don't know. Maybe we could also or instead fix the lib-jitsi-meet.

File: react/features/base/connection/actions.native.js [hunk 1]
```diff
@@ -322,7 +324,20 @@ export function disconnect() {
...
+                    .then(() => {
+                        // If the conference is still in the leaving flag it
+                        // means that for some reason lib-jitsi-meet has not
+                        // dispatched the left event.
+                        const { leaving }
+                            = getState()['features/base/conference'];
+
+                        if (leaving === conference_) {
+                            dispatch(conferenceLeft(conference_));
+                            logger.warn(
```
Comment (lyubomir) [inline react/features/base/connection/actions.native.js]: I don't get the idea of this warn here. So you saw that lib-jitsi-meet doesn't emit the CONFERENCE_LEFT event, we chose to fix it in jitsi-meet, you chose to not fix it in lib-jitsi-meet, and now you're going to warn in jitsi-meet that lib-jitsi-meet didn't do it!? It's already stated in the comment that that's what happened, it's the whole reason why this additional source code exists in jitsi-meet...

**Comment by paweldomas** [comment]
> I think this PR takes the wrong approach. The fix belongs in lib-jisti-meet IMHO.

We were discussing this with Lyubomir and my understanding was that we can have a safeguard in jitsi-meet, but we should probably also fix the lib-jitsi-meet.

> I don't really understand the doReject parameter either, since nobody can set it AFAICT.

You can complain about it to the person who wrote it.

> If the disconnect function is going to pretend to succeed, it should emit fake events (or whatever is needed) so the handling is consistent by the consumers of the API.

Yes, that will be the fix in lib-jitsi-meet about

File: react/features/base/connection/actions.native.js (hunk 1 above)
Comment (paweldomas) [reply react/features/base/connection/actions.native.js]: But I can have it removed if you want

File: react/features/base/connection/actions.native.js [hunk 2]
```diff
@@ -322,7 +324,20 @@ export function disconnect() {
...
+                                'JitsiMeetConference.leave resolved without'
```
[the remark this answers is not in the document]
Comment (paweldomas) [reply react/features/base/connection/actions.native.js]: I will have this fixed if we decide to leave the warn otherwise it will be gone

**Comment by saghul** [comment]
> We were discussing this with Lyubomir and my understanding was that we can have a safeguard in jitsi-meet, but we should probably also fix the lib-jitsi-meet.

Why safeguard against "impossible" situations? Let's fix it and fix it for good.

> You can complain about it to the person who wrote it.

I wasn't complaining to you. I was pointing out that the code you linked had the ability to be correct, but that parameter meant it did the wrong thing IMHO.

> Yes, that will be the fix in lib-jitsi-meet about

Then why not fix it there, bump the ljm dependency and drop this PR? (or am I missing something obvious here?)

**Comment by paweldomas** [comment]
> Then why not fix it there, bump the ljm dependency and drop this PR? (or am I missing something obvious here?)

Yes, we can do that

**Comment by paweldomas** [comment]
> I don't really understand the doReject parameter either, since nobody can set it AFAICT.

Btw. this flag is passed in the timeout function set to true

**Comment by paweldomas** [comment]
It annoying that I'm discussing this issue with @lyubomir then a PR ends up not needed. It should have been dropped in the discussion in the first place

**Comment by paweldomas** [comment]
When looking at lib-jitsi-meet I don't know if it's the best idea to dispatch the CONFERENCE_LEFT there. Even though there is a timeout it can be that some app may be interested to know when the conference has been left for real. In jitsi-meet currently we don't care, but if some other app would, then we would be emitting a false event there saying it's left, but actually the participant is still in the room.

A more correct approach to me would be to not resolve the leave() promise in case of a timeout, but reject it and then we're still emitting CONFERENCE_LEFT, because we've decided to not care.

**Comment by paweldomas** [comment]
I'm opening these 2 PRs as possible options:

https://github.com/jitsi/lib-jitsi-meet/pull/743

https://github.com/jitsi/jitsi-meet/pull/2987

**Pull request closed without merging**